### ISSUE TITLE
BatchNorm channel deletion doesn't work

### ISSUE THREAD

**Issue by s36srini** [Raukk/tf-keras-surgeon#3]
Error: ListWrapper cannot index a list, occurs in surgeon.py: `index[layer.axis] = slice(None)`.

Looking at the code, I'm not sure what's going on exactly - shouldn't this just be a matter of setting the mask to be based on the previous inputs outputmask? It looks like that's being done here, but the above is syntactically incorrect, unless index is supposed to be a numpy array.. but it's not.

**Comment by s36srini** [comment]
I think I might have resolved it by unwrapping the list wrapper and iterating through its indices and setting index[i] = slice(None)

It seems to work fine (dimensional wise), but lmk if there's a more proper work around.

**Comment by Raukk** [comment]
Thanks for reporting this.
I'll look into it. 

What Version of Tensorflow are you running? There were some major changes between (1.13.1 or earlier) and 1.14.0 (or higher). I fixed some locations, but others may still exist (or I may have broken others). Most of them had to do with Tensorflow changing things that were lists (of length one) to single items.

It sounds like you have a potential fix, so, feel free to open a pull request.

Thanks,

PS. Would you mind me asking how you are using pruning (or planning to)?

**Comment by s36srini** [comment]
Yup, sure thing, will open a PR. I'm running TF v 1.18 I believe (nightly). 

I'm currently working on channel wise pruning via elimination through minimum weight channel sum for MobileNet pointwise convolutional layers. 

I'm currently trying to implement support for DepthwiseConv2D (not the seperable version, though I was looking into whether I could apply fusion to save myself some trouble as I noticed that someone already added support for that :'))

For DepthwiseConv2D, for some reason if the outbound_mask is None, it does not correctly dimension the feature map on its own. Perhaps we can talk about this if you're quite familiar with this package? 

BTW: Much thanks for adding a tf.keras version for this package.

**Comment by Raukk** [comment]
Thanks, I'll see if I can find out if this is something related to 1.18/nightly, or just related to the depthwise convolutions.

I have an okay grasp of the theory on depthwise and seperable convolutions, but I have not dug into the code for them, so, I don't know how TF.keras has implemented it.
My gut says you probably don't want to mess with or prune the intermediate value of these types of layers. 
Though, for the depth wise part that mobile nets uses, it seems like it's similar to running a normal 3d convolution of kernel NxNx1 (after adding a 1 length dimension) so that the same filters are applied to each channel seperatly. But however it's quite possible I'm  misunderstanding how this convolutions is working. 
The surgeon code primarily depends on how TF.keras exposes the weights. 


I'm assuming "minimum weight channel sum" rates channels for pruning based on the total magnitude of the weights for that channel?


I'm glad you've found this useful.

**Comment by Raukk** [comment]
Actually, the harder part could be pruning the layer before a depthwise, because the next layer being a depthwise outputs a number of channels that's a multiple of the input channels, and so reducing the number of input channels would also reduce the number of output channels. This would have the effect of basically pruning the depthwise layer at the same time. The pointwise layer would stop this propagation because it has a set number of outputs per conv window, but if multiple depthwise layers are strung together without a layer defining the outputs in-between then the prune would need to flow through them all.

It should be doable, but I think it could get complex.

**Comment by s36srini** [comment]
Yes, for depth wise convolution , each kernel is only applied to a single channel if the depth multiplier is 1 -> the ith kernel is applied to the ith channel. So modifying the inbound mask is just a matter of slicing it based on the outputshape of the depthwiseconv layer (I've recently added support for this). 
I almost have it working except it fails on the last layer, I've posted an issue regarding it, I have no clue why it's giving me trouble when setting the new layer weights, this is the exact same procedure for all convolutional layers... 
Here is the link for that issue: https://github.com/BenWhetton/keras-surgeon/issues/36

**Comment by Raukk** [comment]
Thanks for linking to that issue.

Can you create a really short example; no model training or evaluation, just prune the first channel. 
Maybe have the code loop from the first layer to the last layer, to test each one.


Sudo code (sorry, on my phone)
For layer in layers_to_prune:
    model = delete_channels(model, layer, [0])

**Comment by s36srini** [comment]
Sure, here is the code I'm currently executing:
```

import tfkerassurgeon

model = final_model
surgeon = tfkerassurgeon.Surgeon(model)
        # i must be determined by iterating through weights of layer and figuring out which channels are zeroed
        # final_model = delete_channels(final_model, layer, [i])
for i, w in enumerate(model.get_weights()):
   # print(model.weights[i].name, model.weights[i].shape)
    if re.match("conv_pw_\d+/kernel", model.weights[i].name):
        weights = np.transpose(w[0][0])
        zero_channels = list(filter(lambda x: np.sum(weights[x]) == 0, range(weights.shape[0])))
        print(model.weights[i].name[:-9])
        surgeon.add_job('delete_channels', model.get_layer(name=model.weights[i].name[:-9]), channels=[0])
        #print(len(zero_channels), model.layers[i])l

new_model = surgeon.operate()
```

I have tested keras surgeon with smaller models, and it seems to work just fine. If you'd like you can replicate the above by simply importing Mobilenet. 

`model = tf.keras.applications.MobileNet()`

**Comment by s36srini** [comment]
Update: It was because the actual reshape tensor's target shape wasn't changing accordingly to the pruned input mask even though the outbound mask was being set correctly. The solution is temporary as it only works for squeezable reshapes.

**Comment by Raukk** [comment]
That makes sense. 


A reshape has a defined output shape that is user specified and not necessarily derivative from the Input shape (except that the totals must match). Because of that, we cannot make assumptions about how to change the output shape of the reshape layer when we prune it's input channel. There are probably other layers that are similarly limited. 


The following is probably confusing and not helpful, but it was my attempt to give an example as to why we cannot prune any layers that are followed by a reshape.   

As a weird thought example; imagine something like for an auto-encoder/gan for the MNIST data-set, and make a model with a few convolutions, then a flatten followed by a large dense layer like a standard Lenet, but then added a reshape followed by some de-convolutions.  
While this network is a bit weird, it is fairly close to a common auto-encoder design 
(This probably isn't the best example, but: https://github.com/keras-team/keras/blob/master/examples/variational_autoencoder_deconv.py)
Now, consider; what would happen if you pruned the middle dense layer that is followed by the reshape? Since the Reshape has to put the dense layers output shape back to a grid of channels for the deconvolution layers, how would it know what to reduce, should it remove from the X dimension, the Y dimension, or the channels? And would we even have the right number of outputs?

**Comment by s36srini** [comment]
Yes exactly, I was lucky  because I was dealing with reshape in the context of squeezable tensors, so identifying the dimension that contains channels is easy work since it would just be the last dimension. 

Here is my fork with the changes I have made, it successfully works with MobilenetV1 (I probably need to clean all the prints, so people would actually want to use it :')

https://github.com/s36srini/tf-keras-surgeon
